Lib/functools.py

```diff
@@ -816,8 +815,15 @@ def register(cls, func=None):
         dispatch_cache.clear()
         return func
 
+
     def wrapper(*args, **kw):
-        return dispatch(args[0].__class__)(*args, **kw)
+        # If our decorated function is in the first arg's attributes,
+        # we are using a method.
+        if getattr(args[0], func.__name__, False):
```

Comment:
This looks like a fragile hack. Some arguments may have the same attribute just by coincidence.

Reply:
Hm, yes, I wasn't sure if this was the best solution or not. The main issue is that the decorated function is not an instance of `types.MethodType` when the decorator reaches it. I also looked at frame stack hacks but that would cause conflicts if it was decorating closures. At this point it seems to me a new `methodsingledispatch` is in order, with dispatching on `args[1]`. e.g.:

```python
def methodsingledispatch(func):
     wrapped = singledispatch(func)
     def wrapper(*args, **kwargs):
        wrapped.dispatch(args[1].__class__)(*args, **kw)
     wrapper.register = wrapped.register
     update_wrapper(wrapper, func)
     return wrapper
```
 Alternatively, I think allowing the user to specify which arg index for singledispatch to dispatch on would also be a possible solution (and a nice feature to add in general). Then it would be as simple as `methodsingledispatch = partial(singledispatch, index=1)` or something like that.

Reply:
This is an API change, so this probably should go through python-ideas/python-dev.

Reply:
Okay, I will post to python-dev.